## Les 1

Basisbewerkingen:

In [1]:
# Gehele deling
print(142 // 7)

# Rest bij deling = modulo
print(142 % 7)

# Gekke bewerkingen
getal = 100
getal /= 50
print(getal)

20
2
2.0


+= <br>
-= <br>
*= <br>
/= <br>
%= <br>
//= <br>
**= <br>

In [20]:
# Van karakter naar numerieke waarde
print(ord('a'))

# Van numerieke waarde naar karakter
print(chr(104))
print(chr(ord('a')))

97
h
a


In [23]:
# Vervang spaties door end-of-line characters
tekst = "Dit is een string."
print(tekst.replace(" ", "\n"))

Dit
is
een
string.


String formatting:

In [25]:
leeftijd = 17
hobby = "wielrennen"

# percent methode:
print("De gebruiker is %d jaar oud en doet graag %s." % (leeftijd, hobby))

# format methode:
print("De gebruiker is {0} jaar oud en doet graag {1}.".format(leeftijd, hobby))

# f-string methode:
print(f"De gebruiker is {leeftijd} jaar oud en doet graag {hobby}.")

De gebruiker is 17 jaar oud en doet graag wielrennen.
De gebruiker is 17 jaar oud en doet graag wielrennen.
De gebruiker is 17 jaar oud en doet graag wielrennen.


Wanneer je een string splicet, is de eindwaarde __niet inclusief__.

In [12]:
string = "string"

# Eindwaarde is niet inclusief
print(string[0:2])
print(string[1:2])

print("")

# Laatste letter printen aan de hand van de lengte van de string
print(string[len(string)-1])

# Laaste 3 letters printen
print(string[len(string)-3:len(string)])

st
t

g
ing


## Les 2

In [20]:
# ! Range is niet inclusief voor de eindwaarde.

for i in range(3,6):
    print(i)

3
4
5


Een range is een collectie, naar een string ook. <br>
We kunnen dus loopen op een string:

In [33]:
zin = "Dit is een zin, waarvan we het aantal klinkers gaan bepalen voor het eerste punt. De klinkers die je nu leest zullen dus niet worden meegeteld."
aantalKlinkers = 0

for letter in zin:
    if letter == ".":
        # Met 'break' exiten we uit de for-loop
        break
    if letter in "aeiouy":
        aantalKlinkers+=1

print(aantalKlinkers)


# Of we kunnen dit herschrijven als:

aantalKlinkers = 0
positie = 0

while positie < len(zin) and zin[positie] != ".":
    if zin[positie] in "aeiouy":
        aantalKlinkers += 1
    positie += 1

print(aantalKlinkers)

27
27


Opmaak van print-statements. Rechts uitlijnen:

In [38]:
for n in range(1,10):
    # Beginspacing vóór het eerste getal
    lijn = "%5d" % (n)
    for m in range(2,5):
        # Spacing voor elk nieuw getal op eenzelfde lijn
        lijn += "%10d" % (n**m)
    print(lijn)

    1         1         1         1
    2         4         8        16
    3         9        27        81
    4        16        64       256
    5        25       125       625
    6        36       216      1296
    7        49       343      2401
    8        64       512      4096
    9        81       729      6561


Links uitlijnen kan met _%-5d_

In [39]:
for n in range(1,10):
    # Beginspacing vóór het eerste getal
    lijn = "%-5d" % (n)
    for m in range(2,5):
        # Spacing voor elk nieuw getal op eenzelfde lijn
        lijn += "%-10d" % (n**m)
    print(lijn)

1    1         1         1         
2    4         8         16        
3    9         27        81        
4    16        64        256       
5    25        125       625       
6    36        216       1296      
7    49        343       2401      
8    64        512       4096      
9    81        729       6561      


Lists

In [47]:
lijst = []

lijst.append("item 1")
lijst.append("item 2")

print(lijst)

lijst.remove("item 1")

print(lijst)

['item 1', 'item 2']
['item 2']


Nog wat formatting: de formatting met het dubbele punt is niet inclusief qua lengte.

In [34]:
var1 = "test"
var2 = "test2"

print(f"{var1:10}{var2:10}")

print("{0:10}{1:10}".format(var1, var2))
print("{:10}{:10}".format(var1, var2))

print("%-9s %-9s" % (var1, var2))

print("%10.2f" % 3.14151324234324)

test      test2     
test      test2     
test      test2     
test      test2    
      3.14


In [7]:
dagen = ['Ma', 'Di', 'Woe', 'Do', 'Vr', 'Za', 'Zo']

# Met end character van print
for dag in dagen:
    print("%4s" % dag, end="")

print("")
print("")

# Met percent
eersteRij = ""
for dag in dagen:
    eersteRij += "%4s" % dag
print(eersteRij)

print("")

#! Met f string
kalender = ""
for dag in dagen:
    kalender += f'{dag:>4}'
print(kalender)

  Ma  Di Woe  Do  Vr  Za  Zo

  Ma  Di Woe  Do  Vr  Za  Zo

  Ma  Di Woe  Do  Vr  Za  Zo


__Referentiesemantiek__ bij lijsten: <br>
Als we twee lijsten gelijkstellen en we passen iets aan in de ene, zal het ook veranderen in de andere. <br>
Dit kunnen we tegengaan door een kopie van de lijst te nemen met 'list'.

In [89]:
# BELANGRIJK!
lijst = ["item 1", "item 2", "item 3", "item 4"]
gelinkteLijst = lijst
gekopieerdeLijst = list(lijst)

print(lijst)
print(gelinkteLijst)
print(gekopieerdeLijst)

print("")
lijst[0] = "AANPASSING"

print(lijst)
print(gelinkteLijst)
print(gekopieerdeLijst)

['item 1', 'item 2', 'item 3', 'item 4']
['item 1', 'item 2', 'item 3', 'item 4']
['item 1', 'item 2', 'item 3', 'item 4']

['AANPASSING', 'item 2', 'item 3', 'item 4']
['AANPASSING', 'item 2', 'item 3', 'item 4']
['item 1', 'item 2', 'item 3', 'item 4']


String zijn <u>__IMMUTABLE__</u>!!
Op een string kunnen we __geen replacement van een bepaald karakter__ uitvoeren:

In [ ]:
s = "Dit is een string"
print(s[3])
# Dit geeft 't' terug

s[3] = 's'
print(s)
# Error

Logica om aan de user dingen te blijven vragen:

In [ ]:
lijst = []
volgende = input("Eerste user-input: ")
while volgende != "":
    lijst.append(volgende)
    volgende = input("Volgende inputs")

Achterwaards indexeren bij lijsten:

In [95]:
lijst = [1,2,3,4,5]

# Begin helemaal achteraan met index -1
print(lijst[-1])
# De rest is gewoon telkens 1 minder
print(lijst[-2])

5
4


Bijvoorbeeld om te checken of iets een palindroom is:

In [99]:
palindromaleLijst = ["L", "E", "P", "E", "L"]
nietPalindromaleLijst = ["L","I","J","S","T"]

def checkPalindroom(lijst):
    isPalindroom = True
    for r in range(len(lijst)//2):
        if lijst[r] != lijst[-(r+1)]:
            isPalindroom = False
            break
    return isPalindroom

print(palindromaleLijst, checkPalindroom(palindromaleLijst))
print(nietPalindromaleLijst, checkPalindroom(nietPalindromaleLijst))

['L', 'E', 'P', 'E', 'L'] True
['L', 'I', 'J', 'S', 'T'] False


## Les 3

Sorteerfunctie voor een lijst:

In [4]:
lijst = [4,9,1,3,5,2]

# Overloop elke plaats van de lijst en kijk of er een ander element kleiner is
# Als er een ander element kleiner is, verwissel dan het huidige element en het kleinere element

#! range moet niet tot het laatste element, aangezien dit element sowieso goed staat
# range kan dus lopen tot len(lijst)-1 ipv gewoon len(lijst)

for goed in range(len(lijst)-1):

    minimum = goed

    # Overloop de andere elementen van de lijst
    #! We starten dus op index goed+1
    for i in range(goed+1,len(lijst)):
        if lijst[i] < lijst[minimum]:
            minimum = i

    if goed != minimum:
        tmp = lijst[goed]
        lijst[goed] = lijst[minimum]
        lijst[minimum] = tmp

print(lijst)

[1, 2, 3, 4, 5, 9]


In [8]:
import math

# In een lijst kunnen we eender wat voor object zetten.
lijst = ['muis', 42, 3.14, math.pi, ["een", "twee"]]

#! Een kopie maken, want REFERENTIESEMANTIEK
kopie = list(lijst)

In [29]:
# Een lege lijst maken
legeLijst = list()
print("Lege lijst: ", legeLijst)

# Een lijst met een bepaald aantal keer hetzelfde item
nullen = [0]*5
print("Nullen: ", nullen)

a = ["1","2"]
b = ["3", "4"]

# Lijsten samenvoegen
opgeteld = a + b
print("Opgeteld: ", opgeteld)

#! Dit kan ook met extend
# a wordt dan ge-extend en b blijft gelijk
a.extend(b)
print("Extended: ", a)

Lege lijst:  []
Nullen:  [0, 0, 0, 0, 0]
Opgeteld:  ['1', '2', '3', '4']
Extended:  ['1', '2', '3', '4']


In [11]:
# Toevoegen met append()
lijst.append("toegevoegd item")
print(lijst)

# Verwijderen met remove()
lijst.remove(math.pi)
print(lijst)

['muis', 42, 3.14, 3.141592653589793, ['een', 'twee'], 'toegevoegd item', 'toegevoegd item']
['muis', 42, 3.14, ['een', 'twee'], 'toegevoegd item', 'toegevoegd item']


Geef aan de method _insert_ mee __vóór__ welke index je wil inserten!

In [95]:
lijst = [0,1,2,3,4]

# Items inserten op een bepaalde plaats
#! Er zal een item toegevoegd worden vóór de index in het argument
lijst.insert(2,"-")

print(lijst)

[0, 1, '-', 2, 3, 4]


In [96]:
# Een lijst kunnen we helemaal leegmaken:
lijst.clear()
print(lijst)

[]


In [17]:
lijst = [0,1,2,3,4]

# Indien pop geen argument krijgt, zal hij het laatste element verwijderen
lijst.pop()
print("Pop zonder argument geeft: ", lijst)

lijst.pop(0)
print("Pop het eerste element (index 0): ", lijst)

Pop zonder argument geeft:  [0, 1, 2, 3]
Pop het eerste element (index 0):  [1, 2, 3]


In [32]:
# Een lijst heeft een aantal functies

lijst = [0,1,2,3,4]

print("len: ", len(lijst))
print("max: ", max(lijst))
print("min: ", min(lijst))
print("sum: ", sum(lijst))

print("")

#! Min en max werken ook met strings, op basis van ASCII waarde
lijstMetStrings = ["abc", "bcd", "cde"]
print("min string: ", min(lijstMetStrings))
print("max string: ", max(lijstMetStrings))

len:  5
max:  4
min:  0
sum:  10

min string:  abc
max string:  cde


Strings zijn __immutable__.
Lists hebben __referentiesemantiek__.

In [24]:
# Immutable
s1 = "Deze string is immutable"
s2 = s1
s2 += "."
print(s1)
print(s2)

# Referentiesemantiek, beide lijsten worden aangepast!
l1 = [2,3,6,20, 168, 7581]
l2 = l1
l1 += [7828354]
print(l1)
print(l2)

Deze string is immutable
Deze string is immutable.
[2, 3, 6, 20, 168, 7581, 7828354]
[2, 3, 6, 20, 168, 7581, 7828354]


In [35]:
# Slicen met NIET-INCLUSIEVE eindindex
lijst = [2,3,6,20,168,1729,7581,7828354]
print(lijst[1:4])
print(lijst[0:4])
print(lijst[:4])
print(lijst[6:len(lijst)])
print(lijst[6:])
print(lijst[6:-3])

[3, 6, 20]
[2, 3, 6, 20]
[2, 3, 6, 20]
[7581, 7828354]
[7581, 7828354]
[]


In [37]:
lijst = [2,3,6,20,168,1729,7581,7828354]

print(lijst)
# We voegen het getal 14 toe vóór index 3 (20)
lijst.insert(3,14)
print(lijst)

#! We halen de index van een waarde in de lijst op
plaatsVan168 = lijst.index(168)
print("Index van het gal 168:", plaatsVan168)

[2, 3, 6, 20, 168, 1729, 7581, 7828354]
[2, 3, 6, 14, 20, 168, 1729, 7581, 7828354]
Index van het gal 168: 5


In [39]:
# Twee verschillende plaatsen in het geheugen
l1 = [2,3,6,20,168,1729,7581,7828354]
l2 = [2,3,6,20,168,1729,7581,7828354]

#! Checken of twee lijsten effectief gelijk zijn met IS
gelijk = l1 is l2
print("Twee verschillende plaatsen in het geheugen: ",gelijk)


# Zelfde plaats in geheugen en zelfde referentie
l1 = [2,3,6,20,168,1729,7581,7828354]
l2 = l1

gelijk = l1 is l2
print("Zelfde plaats in geheugen en zelfde referentie: ",gelijk)

Twee verschillende plaatsen in het geheugen:  False
Zelfde plaats in geheugen en zelfde referentie:  True


In [43]:
lijst = [6,3,9,2,5,1,0]

# De functie sorted returnt de gesorteerde lijst, string, tupel, set, ...
print("sorted(): ", sorted(lijst))
print("De lijst blijft onveranderd:  ", lijst)

# De method sorteert de effectieve lijst en past hem ook aan
lijst.sort()
print("De aangepaste lijst: ", lijst)

sorted():  [0, 1, 2, 3, 5, 6, 9]
De lijst blijft onveranderd:   [6, 3, 9, 2, 5, 1, 0]
De aangepaste lijst:  [0, 1, 2, 3, 5, 6, 9]


Sorted() werkt voor __eender welk object!__

Speciale prints:

In [49]:
matrix = []
matrix.append([1,0,0,0])
matrix.append([0,1,0,0])
matrix.append([0,0,1,0])
matrix.append([0,0,0,1])


for line in matrix:
    for element in line:
        #! We willen geen nieuwe lijn karakter
        #! We voegen een TAB toe
        print(element, end='\t')
    # Dit is de standaard, namelijk een nieuwe lijn karakter \n    
    print("", end="\n")

1	0	0	0	
0	1	0	0	
0	0	1	0	
0	0	0	1	


__TUPELS__ = 'immutable lists'

In [46]:
tupel = ("immutable", "list")
tupelZonderHaakjes = "zonder", "haakjes", "werkt", "ook", 69

print(tupel)
print(tupelZonderHaakjes)


('immutable', 'list')
('zonder', 'haakjes', 'werkt', 'ook', 69)


Een tupel is immutable, maar de inhoud niet. We kunnen dus een nieuwe tupel aanmaken obv twee objecten die niet immutable zijn en die objecten vervolgens aanpassen.

In [59]:
a = ["Lists"]
b = "Immutable string. "

tupel = (a,b)
print(tupel)

#! a is een list. Lists zijn niet immutable, dus we kunnen a aanpassen.
a += ["zijn niet immutable"]
#! b is een string en is dus immutable. We kunnen de string niet aanpassen.
b += "Deze aanpassing wordt niet getoond in de tupel."
print(tupel)

print(b)


(['Lists'], 'Immutable string. ')
(['Lists', 'zijn niet immutable'], 'Immutable string. ')
Immutable string. Deze aanpassing wordt niet getoond in de tupel.


Verzamelingen = __SETS__ <br>
De items hebben geen specifieke volgorde.

Bij sets kunnen we __geen items opvragen aan de hand van een index__ want er is geen specifieke volgorde. `set[1]` ofzo werkt dus niet.

In [81]:
studenten = {"Jan","Piet","Joris","Korneel"}
print(studenten)

#! Add in plaats van append
studenten.add("Carolien")
print(studenten)

# Remove is hetzelfde als bij lists
studenten.remove("Piet")
print(studenten)

#! Elementen in sets zijn uniek dus we kunnen 'Korneel' geen tweede keer toevoegen
studenten.add("Korneel")
print(studenten)

{'Piet', 'Korneel', 'Joris', 'Jan'}
{'Joris', 'Jan', 'Carolien', 'Piet', 'Korneel'}
{'Joris', 'Jan', 'Carolien', 'Korneel'}
{'Joris', 'Jan', 'Carolien', 'Korneel'}


In [103]:
A = {"A","B","C","D"}
B = {"B","E"}

unie = A | B
unie = A.union(B)
print("unie: \t\t", unie)

doorsnede = A & B
doorsnede = A.intersection(B)
print("doorsnede: \t", doorsnede)

verschil1 = A - B
verschil1 = A.difference(B)
print("verschil1: \t", verschil1)

verschil2 = B - A
verschil2 = B.difference(A)
print("verschil2: \t", verschil2)

unie: 		 {'A', 'B', 'C', 'E', 'D'}
doorsnede: 	 {'B'}
verschil1: 	 {'C', 'A', 'D'}
verschil2: 	 {'E'}


In [94]:
verschil1 = A - B
verschil2 = B - A

# We kunnen ook een item removen van een set
print(verschil2)
verschil2.remove("E")
print(verschil2)

print("")

#! Als we discarden maakt het niet uit dat het element niet effectief in de set zit
print(verschil2)
verschil2.discard("E")
print(verschil2)

print("")

# We discarden een element dat wel bestaat
print(verschil1)
verschil1.discard("C")
print(verschil1)

print("")

#! Of we kunnen heel de set leegmaken
verschil1.clear()
print(verschil1)

{'E'}
set()

set()
set()

{'C', 'A', 'D'}
{'A', 'D'}

set()


__DICTIONARIES__ -> verzamelingen associaties <br>
Keys zijn uniek <br>
Values moeten niet uniek zijn

In [114]:
telefoonboek = {} # een lege dictionary
telefoonboek["jens"] = "7657"
telefoonboek["jarne"] = "7657"
telefoonboek["michiel"] = "2402"

print(telefoonboek)
telefoonboek["michiel"] = "1010"
print(telefoonboek)

print("")

print("Keys: \t", telefoonboek.keys())
print("Values:\t", telefoonboek.values())
print("Items: \t", telefoonboek.items())

{'jens': '7657', 'jarne': '7657', 'michiel': '2402'}
{'jens': '7657', 'jarne': '7657', 'michiel': '1010'}

Keys: 	 dict_keys(['jens', 'jarne', 'michiel'])
Values:	 dict_values(['7657', '7657', '1010'])
Items: 	 dict_items([('jens', '7657'), ('jarne', '7657'), ('michiel', '1010')])


In [110]:
telefoonboek = {"jens":400, "jarne":7657, "michiel":2402}
legeD = dict()
print(legeD)
kopie = dict(telefoonboek)
print(kopie)

# dictionary[key] en dictionary.get(key) returnen allebei de value van de key
print(kopie['jarne'])
print(kopie.get("jarne"))

#! We halen het vorige item op aan een bepaalde key
print(kopie.get("jarne",-1))

# We verwijderen een key-value-paar
print(kopie.pop('jens'))
print(kopie)

#! We kunnen niet zomaar poppen zonder argument !
# kopie.pop()

{}
{'jens': 400, 'jarne': 7657, 'michiel': 2402}
7657
7657
7657
400
{'jarne': 7657, 'michiel': 2402}


Sorted() werkt dus zoals eerder vermeld ook op dictionaries.

In [2]:
telefoonboek = {"jens":400, "jarne":7657, "michiel":2402}

#! sorted(dictionary) returnt de gesorteerde KEYS !
print(sorted(telefoonboek))

#! sorted(dictionary).items returnt de gesorteerde ITEMS !
print(sorted(telefoonboek.items()))

['jarne', 'jens', 'michiel']
[('jarne', 7657), ('jens', 400), ('michiel', 2402)]


In [5]:
# Itereren over dictionaries
#! Gebruik telefoonboek.items()
for (key, value) in telefoonboek.items():

    print("key:", key, end=", ")
    print("value:", value)

print("")

# Dit komt op hetzelfde neer:
for tupel in telefoonboek.items():

    print("key:", tupel[0], end=", ")
    print("value:", tupel[1])

# Itereren over de keys kan gewoon zo:
for key in telefoonboek:
    print("key:", key)

key: jens, value: 400
key: jarne, value: 7657
key: michiel, value: 2402

key: jens, value: 400
key: jarne, value: 7657
key: michiel, value: 2402
key: jens
key: jarne
key: michiel


In [8]:
#Frequentietabel

tekst = input("Enter text: ")
telling = {}

for c in tekst:
    # De letter komt nog niet voor, we zetten de telling van de letter op 1
    if c not in telling:
        telling[c] = 1
    # De letter zit wel al in de telling en we verhogen de telling van de letter met 1
    else:
        telling[c] += 1

print(telling)
print(sorted(telling))

# We printen alle letters en hun telling die letters zijn (geen symbolen en cijfers)
for c in sorted(telling):
    if 'a' <= c.lower() <= 'z':
        print(c,telling[c])

{'D': 1, 'i': 2, 't': 3, ' ': 3, 's': 2, 'e': 3, 'n': 1}
[' ', 'D', 'e', 'i', 'n', 's', 't']
D 1
e 3
i 2
n 1
s 2
t 3


We kunnen .get() een default waarde meegeven voor het geval de key niet bestaat.

In [3]:
legeDict = {}
print(legeDict.get("Niet bestaande key"),0)

None 0


De methode `.keys()` en `.values()` geven een keyset en valueset terug en geen gewone lijst of set. We kunnen deze in een for-lus gebruiken zonder problemen, maar de gewone lijst/set operaties werken niet. Om dat op te lossen gebruiken we altijd `set(dict.keys())` en `list(dict.values())`. 

## Les 4

In [1]:
def schikTekst(t):
    res = "="*42 + '\n'
    geschikt = 0
    lijnpositie = 0
    while geschikt < len(t):
        if lijnpositie >= 42 and t[geschikt] == ' ':
            res += "\n"
            lijnpositie = 0
        res += t[geschikt]
        geschikt += 1
        lijnpositie += 1
    return res + '\n'+ "="*42

print(schikTekst(input("Geef een tekst in: ")))

# Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas gravida nisi sed mattis vulputate. Etiam non pulvinar tortor. Fusce bibendum condimentum lectus. Integer tempus ante eget mauris volutpat iaculis. Maecenas nec massa varius, convallis purus pharetra, efficitur purus. Phasellus et lobortis enim. Fusce quam nisl, sodales non est aliquam, tempus vehicula odio. Vestibulum eget ipsum nibh.

Lorem ipsum dolor sit amet, consectetur adipiscing
 elit. Maecenas gravida nisi sed mattis vulputate.
 Etiam non pulvinar tortor. Fusce bibendum
 condimentum lectus. Integer tempus ante eget
 mauris volutpat iaculis. Maecenas nec massa
 varius, convallis purus pharetra, efficitur
 purus. Phasellus et lobortis enim. Fusce quam
 nisl, sodales non est aliquam, tempus vehicula
 odio. Vestibulum eget ipsum nibh.


Default waarden meegeven voor de parameters van een functie. <br>
Dit kan __enkel met constanten, dus geen variabelen!__

In [12]:
def testFunctie(parameter1, parameter2 = "Default"):
    return parameter1 + parameter2

print(testFunctie("Test"))

TestDefault


Documenteren van functies:

De  `%128` of modulo 128 zorgt ervoor dat je terug in het begin van de tabel komt. <br>
Neem nu bv dat we `125+5=130` hebben. Door de modulo 128 te pakken, komen we terug in het begin van de ASCII tabel terecht bij 2.

In [36]:
def codeer(tekst,k):
    '''
        Gebruik de sleutel k om de tekst te coderen.
        Elk teken wordt vervangen door het teken waarvan
        de ordinaat k hoger is (modulo 128).

        Parameters
        ----------
        tekst : str
            de te coderen tekst
        k : int
            de sleutelwaarde

        Returns
        -------
        str
            de gecodeerde string
        '''
    result = ""
    for teken in tekst:
        result += chr((ord(teken) + k) % 128)
    return result

print(codeer("Aha! Zwart is de kleur.",5))

def decodeer(tekst,k):
    '''
    Gebruik de sleutel k om de tekst te decoderen.
    Elk teken wordt vervangen door het teken waarvan
    de ordinaat k lager is (modulo 128).

    Parameters
    ----------
    tekst : str
        de te decoderen tekst
    k : int
        de sleutel

    Returns
    -------
    str
        de gedecodeerde tekst
    '''
    result = ""
    for teken in tekst:
        result += chr((ord(teken) - k) % 128)
    return result

print(decodeer("Fmf&%_|fwy%nx%ij%pqjzw3",5))

def decodeerBIS(tekst,k):
    '''
    Gebruik de sleutel k om de tekst te decoderen.
    Elk teken wordt vervangen door het teken waarvan
    de ordinaat k lager is (modulo 128).

    Parameters
    ----------
    tekst : str
        de te decoderen tekst
    k : int
        de sleutel

    Returns
    -------
    str
        de gedecodeerde tekst
    '''
    return codeer(tekst,-k)

print(decodeerBIS("Fmf&%_|fwy%nx%ij%pqjzw3",5))

print()
help(decodeerBIS)

Fmf&%_|fwy%nx%ij%pqjzw3
Aha! Zwart is de kleur.
Aha! Zwart is de kleur.

Help on function decodeerBIS in module __main__:

decodeerBIS(tekst, k)
    Gebruik de sleutel k om de tekst te decoderen.
    Elk teken wordt vervangen door het teken waarvan
    de ordinaat k lager is (modulo 128).
    
    Parameters
    ----------
    tekst : str
        de te decoderen tekst
    k : int
        de sleutel
    
    Returns
    -------
    str
        de gedecodeerde tekst



Of een andere manier om te documenteren:

In [27]:
##
# Gebruik de sleutel k om de tekst de coderen.
# Elk teken wordt vervangen door het taken waarvan
# de ordinaat k hoger is.
#
# @param tekst (str): de te coderen tekst
# @param k (int): de sleutel
# @return: de gecodeerde string
#
def codeer(tekst, k):
    result = ""
    for teken in tekst:
        result += chr(ord(teken)+k)
    return result

print(help(codeer))

Help on function codeer in module __main__:

codeer(tekst, k)
    ##
    # Gebruik de sleutel k om de tekst de coderen.
    # Elk teken wordt vervangen door het taken waarvan
    # de ordinaat k hoger is.
    #
    # @param tekst (str): de te coderen tekst
    # @param k (int): de sleutel
    # @return: de gecodeerde string

None


Grootste gemene deler:

In [32]:
def delers(n):
    # 1 is sowieso een deler
    result = [1]
    # range is niet inclusief voor de eindwaarde
    for d in range(2,n+1):
        if n % d == 0:
            result.append(d)
    return result

def ggdSuperNaief(a,b):

    dA = delers(a)
    dB = delers(b)
    grootste = 1
    for ene in dA:
        for andere in dB:
            if ene == andere:
                if ene > grootste:
                    grootste = ene
    return grootste

def ggdSets(a,b):
    dA = set(delers(a))
    dB = set(delers(b))
    return max(dA.intersection(dB))

def ggd(a,b):
    while a != b:
        if a < b:
            b -= a
        else:
            a -= b
    return a

#! Greatest common divisor Euclides
def ggdStappend(a,b):
    if a > b:
        h = a
        a = b
        b = h
    while b % a != 0:
        r = b % a
        b = a
        a = r
    return a

print(delers(210))
print(delers(546))
print(ggdSuperNaief(210,546))

print(ggdSets(210,546))

print(ggd(210,546))

print(ggdStappend(210,546))

[1, 2, 3, 5, 6, 7, 10, 14, 15, 21, 30, 35, 42, 70, 105, 210]
[1, 2, 3, 6, 7, 13, 14, 21, 26, 39, 42, 78, 91, 182, 273, 546]
42
42
42
42


____
!! __PARAMETERS AANPASSEN BINNEN EEN FUNCTIE:__ !! <br><br>
Het is bad practice om parameters van het type `list` aan te passen in de functie. <br><br>
Alles wat iets of wat complex is, wordt niet zomaar opgeslagen in het geheugen, maar bevat een referentie naar andere plaatsen in het geheugen. Ze gebruiken dus referentiesemantiek: (bv lijsten) <br><br>
Als je als parameters voor `x` en `y` twee getallen `a` en `b` meegeeft (of booleans), dan wordt de waarde van `a` en `b` gekopieerd naar de geheugenplaatsen voor `x` en `y`. We kunnen in onze functie dus gewoon de waarden van de parameters aanpassen. <br><br>
Dit is anders bij lijsten. Python zal voor parameters van het type `list` de __referentie__ naar de lijst kopiëren en we mogen de parameter dus niet aanpassen in onze functie. <br><br>
______

Functies moeten altijd opgeroepen worden na hun definitie. <br>
Als de functie echter aangeroepen wordt in de hoofdfunctie `main()`, die gedefiniëerd staat vóór de definitie van de functie, gaat het ook werken:

In [28]:
def main():
    print("%10s" % functie())

def functie():
    return "Test"

main()

      Test


Een functie kan maar 1 waarde teruggeven. Als je dus `return a,b` typt, zal dit overeenkomen met `return (a,b)`, ofwel het returnen van de tupel (a,b).

We kunnen code importeren:

In [ ]:
#Maak een nieuw een bestand in een variabele
#Aangezien we in het bestand zullen schrijven, geven we dit aan met een "w" ("write")
nieuw_bestand = open("bestand.txt","w")

#schrijf een lijn met erachter een enter
nieuw_bestand.write("Dit is de enige zin!\n")

#sluit het bestand
nieuw_bestand.close()

#Open hetzelfde bestand met "w" en probeer een zin toe te voegen
bestand = open("bestand.txt","w")

#schrijf een lijn met erachter een enter
bestand.write("Fout, dit is de enige zin.\n")

#sluit het bestand
bestand.close()

#Open een bestaand bestand in in een variabele
#Aangezien we enkel het bestand zullen lezen, geven we dit aan met een "r" ("read")
bestaand_bestand = open("bestand.txt","r")

#Itereer over elke lijn in het het bestand
for lijn in bestaand_bestand.readlines():
    print(lijn)
    #Itereer over elk woord in elke lijn in het het bestand
    for woord in lijn.split():
        print(woord)
    

#sluit het bestand na het lezen
bestaand_bestand.close()

#Open een bestaand bestand in een variabele
#Aangezien we aan dit bestand zullen toevoegen, geven we dit aan met een "a" ("append")
bestaand_bestand = open("bestand.txt","a")

#schrijf een lijn op het einde van het bestand met erachter een enter
bestaand_bestand.write("Dit is de laatse zin.\n")

#sluit het bestand
bestaand_bestand.close()

bestaand_bestand = open("bestand.txt","r")

for lijn in bestaand_bestand.readlines():
    print(lijn)

bestaand_bestand.close()